src/third_party/field_5x52_asm_impl_cryptopt.h

```diff
@@ -0,0 +1,367 @@
+      "add    %%rbx,%%r14\n"
+      "adc    $0x0,%%rdi\n"
+      "mov    (%%rax),%%rdx\n"
+      "mulx   0x8(%%rsi),%%r11,%%r10\n"
+      "mov    (%%rsi),%%rdx\n"
+      "mulx   0x8(%%rax),%%r9,%%rbx\n"
+      "mov    0x10(%%rsi),%%rdx\n"
+      "mov    %%r15,-0x48(%%rsp)\n"
+      "mov    %%rdi,-0x40(%%rsp)\n"
+      "mulx   (%%rax),%%r15,%%rdi\n"
+      "xor    %%rdx,%%rdx\n"
+      "adox   %%r9,%%r11\n"
+      "adox   %%r10,%%rbx\n"
+      "adcx   %%rbp,%%r13\n"
+      "adcx   %%r8,%%rcx\n"
+      "mov    -0x40(%%rsp),%%rbp\n"
+      "mov    %%r14,%%r8\n"
+      "shrd   $0x34,%%rbp,%%r8\n"
+      "mov    0x8(%%rax),%%rdx\n"
+      "mulx   0x8(%%rsi),%%rbp,%%r10\n"
+      "mov    %%r13,%%rdx\n"
+      "shrd   $0x34,%%rcx,%%rdx\n"
+      "xor    %%r9,%%r9\n"
+      "adox   %%r11,%%rdx\n"
+      "adox   %%r9,%%rbx\n"
+      "mov    %%rdx,%%r11\n"
+      "mov    0x18(%%rsi),%%rdx\n"
+      "mulx   0x20(%%rax),%%rcx,%%r9\n"
+      "adcx   %%rbp,%%r15\n"
+      "adcx   %%rdi,%%r10\n"
+      "mov    0x18(%%rax),%%rdx\n"
+      "mulx   0x20(%%rsi),%%rdi,%%rbp\n"
+      "xor    %%rdx,%%rdx\n"
+      "adox   %%rcx,%%rdi\n"
+      "adox   %%rbp,%%r9\n"
+      "adcx   %%rdi,%%r8\n"
+      "adc    $0x0,%%r9\n"
+      "movabs $0xfffffffffffff,%%rcx\n"
+      "and    %%rcx,%%r14\n"
+      "mov    0x10(%%rax),%%rdx\n"
+      "mulx   (%%rsi),%%rbp,%%rdi\n"
+      "adox   %%rbp,%%r15\n"
+      "adox   %%r10,%%rdi\n"
+      "movabs $0x1000003d10,%%rdx\n"
+      "mulx   %%r14,%%r10,%%rbp\n"
+      "adcx   %%r11,%%r10\n"
+      "adcx   %%rbp,%%rbx\n"
+      "mov    %%r10,%%r11\n"
+      "shrd   $0x34,%%rbx,%%r11\n"
+      "and    %%rcx,%%r10\n"
+      "mulx   %%r8,%%r14,%%rbp\n"
+      "movabs $0x1000003d10000,%%r8\n"
+      "mov    %%r9,%%rdx\n"
+      "mulx   %%r8,%%r9,%%rbx\n"
+      "adox   %%r15,%%r11\n"
+      "mov    $0x0,%%edx\n"
+      "adox   %%rdx,%%rdi\n"
+      "adcx   %%r11,%%r14\n"
+      "adcx   %%rbp,%%rdi\n"
+      "mov    %%r14,%%r15\n"
+      "shrd   $0x34,%%rdi,%%r15\n"
+      "mov    -0x50(%%rsp),%%rbp\n"
+      "mov    %%r10,0x8(%%rbp)\n"
+      "and    %%rcx,%%r12\n"
+      "lea    (%%r12,%%r15,1),%%r12\n"
+      "adox   %%r12,%%r9\n"
+      "adox   %%rdx,%%rbx\n"
+      "and    %%rcx,%%r13\n"
+      "mov    %%r13,0x0(%%rbp)\n"
+      "and    %%rcx,%%r14\n"
+      "mov    %%r9,%%r10\n"
+      "shrd   $0x34,%%rbx,%%r10\n"
+      "mov    %%r14,0x10(%%rbp)\n"
+      "add    -0x48(%%rsp),%%r10\n"
+      "and    %%rcx,%%r9\n"
+      "mov    %%r10,0x20(%%rbp)\n"
+      "mov    %%r9,0x18(%%rbp)\n"
+      "mov    -0x80(%%rsp),%%rbx\n"
+      "mov    -0x78(%%rsp),%%rbp\n"
+      "mov    -0x70(%%rsp),%%r12\n"
+      "mov    -0x68(%%rsp),%%r13\n"
+      "mov    -0x60(%%rsp),%%r14\n"
+      "mov    -0x58(%%rsp),%%r15\n"
+      "mov    -0x50(%%rsp),%%rdi\n"
+      "mov    %%rax,%%rdx\n"
+
+      : "=m"(r[0]), "=m"(r[1]), "=m"(r[2]), "=m"(r[3]), "=m"(r[4])
+
+      : "D"(r), "S"(a), "d"(b), "m"(a[0]), "m"(b[0]), "m"(a[1]), "m"(b[1]),
```

Comment:
Hm, I see. I think I wanted to do this too, but I wasn't able to because of `%rdx` being special for `mulx`. Then I'd need some way of conditionally emitting instructions in the sense of: "If the parameter is in `%rdx` then save it to something else before doing the `mulx`. Similar with the other registers. If the compiler decides which registers the parameters go into, I don't know in the hard coded parts, which registers to use for the intermediate values. And lastly, the more we change this existing code, the less confident we are of the formal correctness. (On that note, I'd actually prefer intel syntax, as that is the one actually verified by Fiat. But I remember that I couldn't get that to work either reliably with gcc / clang. Maybe that changed in recent versions though.)